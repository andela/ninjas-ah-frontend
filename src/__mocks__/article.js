export default (Object.article = {
  title: 'my title well done',
  description: 'description of the article: well',
  body: 'body of the article: well'
});
