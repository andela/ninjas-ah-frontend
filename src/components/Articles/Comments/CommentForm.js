import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import TextareaAutosize from 'react-autosize-textarea';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
// import {
//   faThumbsUp,
//   faClock,
//   faPenAlt,
//   faTrash,
//   faCommentAlt
// } from '@fortawesome/free-solid-svg-icons';
import './Comments.scss';
import { createComment } from '../../../actions';

export class CommentForm extends Component {
  state = { loading: true, comment: '' };

  componentWillMount() {
    const { comments } = this.props;
    // createComment(this.props.slug);
    this.setState({ comments });
  }

  componentWillReceiveProps(nextProps) {
    const { errors, comments, message } = nextProps;
    console.log('will', message, comments);
    if (errors) {
      this.setState({ errors });
    }
    this.setState({ loaded: true });
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = async (e) => {
    e.preventDefault();
    const { isAuth, slug } = this.props;
    if (isAuth) {
      await this.props.createComment({ comment: this.state.comment, slug });
    } else {
      this.setState({ errors: { token: 'Failed to authenticate token' } });
    }
  };

  render() {
    const { isAuth, profile, slug, loading } = this.props;
    const { errors, comment, message } = this.state;
    console.log('loading', loading, message, errors && errors);
    return (
      <div id="comment-form-wrapper">
        {!isAuth || (errors && errors.token) ? (
          <div className="border b-light medium-padding">
            To comment,{' '}
            <Link
              className="bold text-info"
              to={`/login?redirect=articles/${slug}#comment-form-wrapper`}
            >
              Login
            </Link>{' '}
            first
          </div>
        ) : (
          <div className="border b-light medium-padding">
            You are commenting as{' '}
            <Link to="/" className="bold text-success">
              {profile.firstName} {profile.lastName}
            </Link>
          </div>
        )}
        <form onSubmit={this.onSubmit}>
          <div className="input-field">
            <TextareaAutosize
              name="comment"
              onChange={this.onChange}
              value={comment}
              placeholder="Write your comment here..."
              style={{ minHeight: 12 }}
              id="commentBody"
              className="shadow-1"
            />
          </div>
          <div className="input-field">
            <button type="submit" className="button success bold text-white radius-5">
              Comment
            </button>
            {loading ? <span className="medium-padding radius-5">loading</span> : ''}
            {errors && errors.token ? (
              <span className="medium-padding radius-5 text-danger">You need to login first</span>
            ) : (
              ''
            )}
            {errors && errors.message ? (
              <span className="medium-padding radius-5 text-danger">{errors.message}</span>
            ) : (
              ''
            )}
          </div>
        </form>
        <div className="clear" />
      </div>
    );
  }
}

CommentForm.propTypes = {
  createComment: PropTypes.func.isRequired,
  isAuth: PropTypes.bool,
  profile: PropTypes.object,
  comments: PropTypes.array,
  slug: PropTypes.string,
  comment: PropTypes.string,
  rating: PropTypes.number,
  params: PropTypes.object,
  match: PropTypes.object,
  errors: PropTypes.object,
  loading: PropTypes.bool,
  message: PropTypes.object
};

const mapStateToProps = ({
  user: { isAuth, profile },
  comments: { createComment: { comments, message, errors, loading } }
}) => ({ isAuth, profile, comments, message, errors, loading });

export default connect(
  mapStateToProps,
  { createComment }
)(CommentForm);
