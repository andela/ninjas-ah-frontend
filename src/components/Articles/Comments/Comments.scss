@import '../../../assets/css/_colors';
@import '../../../assets/css/_radius';
@import '../../../assets/css/_padding';
@import '../../../assets/css/_margin';
@import '../../../assets/css/_border';

#wrap-comment-thread {
  margin: 0;
  padding: 0;
}
#wrap-comments {
  @extend .border, .b-light;
}

.comments-wrapper {
  margin: 0;
  padding: 0;
}
.comments-wrapper li {
  width: 100%;
  padding: 5px 10px 0 10px;
  position: relative;

  .comment-icon {
    position: relative;
    top: 0px;
    left: 10px;
    float: left;
    width: 10%;
    text-align: center;
    .image {
      margin: 0 auto;
      display: block;
      width: 60px;
      height: 60px;
      overflow: hidden;
      border-radius: 50%;
    }
  }

  .comment-message-wrapper {
    width: 86%;
    position: relative;
    margin-left: 1%;
    float: left;
    padding: 0 1%;
    color: #585858;
    line-height: 30px;
    border-left: 2px solid #cdcdcd;
    button {
      padding: 8px 20px;
      border: none;
      margin-right: 5px;
      cursor: pointer;
      @extend .radius-4;
    }
    .timestamp {
      color: #989898;
    }
    .comment-editor {
      margin: 15px 0 0 0;
      @extend .border, .b-light-grey, .medium-padding, .light, .radius-2;
    }
  }
  .comment-message-wrapper::before {
    position: absolute;
    content: '';
    top: 10px;
    transform: rotate(-90deg);
    left: -18px;
    height: 0;
    border-left: 10px solid transparent;
    border-right: 10px solid transparent;
    border-bottom: 15px solid #cdcdcd;
  }
}

#comment-form-wrapper {
  textarea {
    font-size: 16px;
  }
}
