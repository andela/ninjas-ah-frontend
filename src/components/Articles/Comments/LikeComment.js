import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faThumbsUp } from '@fortawesome/free-solid-svg-icons';
import { likeComment, getLikesComment } from '../../../actions';

export class LikeComment extends Component {
  state = {
    likeClicked: false,
    hasLiked: false,
    likes: 0
  };

  componentWillReceiveProps = (nextProps) => {
    const { comment, commentLikes, profile } = nextProps;
    (commentLikes || []).map(commentLike => (comment.id === parseInt(commentLike.commentId, 10)
      ? this.setState(prevState => ({
        ...prevState,
        likes: commentLike.number,
        hasLiked: commentLike.whoLiked.includes(profile.id)
      }))
      : ''));
  };

  componentWillMount = () => {
    const { comment, getLikesComment } = this.props;
    getLikesComment(comment);
  };

  handleCommentLike = (comment) => {
    const { likeComment } = this.props;
    likeComment(comment, 'like');
  };

  render() {
    const { likeClicked, likes, hasLiked } = this.state;
    const { comment } = this.props;

    return (
      <button
        className={likeClicked ? 'light-red text-black' : 'light text-black'}
        onClick={() => this.handleCommentLike(comment)}
      >
        <FontAwesomeIcon
          icon={faThumbsUp}
          className={hasLiked ? 'text-danger' : 'text-grey'}
          size="1x"
        />
        &nbsp; {likes <= 1 ? <span>{likes} like</span> : <span>{likes} likes</span>}
      </button>
    );
  }
}

LikeComment.defaultProps = { whoLiked: [], profile: {} };

LikeComment.propTypes = {
  comment: PropTypes.object.isRequired,
  likeComment: PropTypes.func.isRequired,
  getLikesComment: PropTypes.func.isRequired,
  numberOfLikes: PropTypes.number,
  commentLikes: PropTypes.array,
  whoLiked: PropTypes.array,
  profile: PropTypes.object
};
const mapStateToProps = ({
  user: { isAuth, profile },
  comments: {
    deleteComment: { message, deleted },
    fetchComments: { comments, errors },
    allComments,
    commentLikes
  }
}) => ({
  isAuth,
  profile,
  comments,
  message,
  errors,
  deleted,
  allComments,
  commentLikes
});

export default connect(
  mapStateToProps,
  { likeComment, getLikesComment }
)(LikeComment);
