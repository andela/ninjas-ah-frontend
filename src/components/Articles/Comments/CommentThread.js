import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { BeatLoader } from 'react-spinners';
import { Link } from 'react-router-dom';
import TextareaAutosize from 'react-autosize-textarea';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faThumbsUp,
  faClock,
  faPenAlt,
  faTrash,
  faCommentAlt
} from '@fortawesome/free-solid-svg-icons';
import './Comments.scss';
import { css } from '@emotion/core';
import { fetchComments, deleteComment, editComment } from '../../../actions';
import { Img } from '../../common';
import timeStamp from '../../../helpers/timeStamp';
import avatar from '../../../assets/images/profile_plaholder.png';

export class CommentThread extends Component {
  state = { comments: [], slug: '', body: '', commentEditor: {}, newComments: {} };

  componentDidMount() {
    const { fetchComments } = this.props;
    fetchComments(this.props.slug);
  }

  componentWillReceiveProps = (nextProps) => {
    const { comments, deleted } = nextProps;
    const { commentKey } = this.state;
    if (comments) {
      this.setState({ comments });
    }
    if (deleted && commentKey !== undefined) {
      delete comments[commentKey];
      this.setState({ comments });
    }
  };

  onDeleteComment = (id, commentKey) => {
    const { slug, deleteComment } = this.props;
    deleteComment({ slug, id });
    this.setState({ id, commentKey });
  };

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onChangeEditComment = (e) => {
    this.setState({ newComments: { [e.target.name]: { value: e.target.value } } });
  };

  toggleCommentEditor = (key) => {
    const { commentEditor } = this.state;
    this.setState(prevState => ({
      ...prevState,
      commentEditor: { [key]: { display: commentEditor[key] && commentEditor[key].display === 'block' ? 'hide' : 'block' } }
    }));
  };

  closeCommentEditor = (e, key) => {
    e.preventDefault();
    const { commentEditor } = this.state;
    this.setState(prevState => ({
      ...prevState,
      commentEditor: { [key]: { display: commentEditor[key] && commentEditor[key].display === 'block' ? 'hide' : 'block' } }
    }));
  };

  onSubmit = async (e, key, id, userId) => {
    e.preventDefault();
    this.setState({ errors: {} });
    const { isAuth, slug, editComment, profile } = this.props;
    const { newComments, comments } = this.state;
    if (isAuth) {
      if (userId === profile.id) {
        if (newComments && Object.keys(newComments).length) {
          const comment = {
            body: newComments[`comment-${key}`].value,
            slug,
            id,
            userId
          };
          await editComment(comment);
          comments[key].body = newComments[`comment-${key}`].value;
          this.setState({ comments, errors: { message: 'You have edited this comment' } });
        } else {
          this.setState({ errors: { message: 'To submit, you need to edit first' } });
        }
      } else {
        this.setState({ errors: { message: 'You can not edit a comment you did not created' } });
      }
    } else {
      this.setState({ errors: { token: 'Failed to authenticate token' } });
    }
  };

  render() {
    const { loading, isAuth, profile } = this.props;
    const { comments, errors, commentEditor, newComments } = this.state;

    return (
      <div id="wrap-comment-thread">
        {loading ? (
          <BeatLoader
            css={css`
              display: block;
              margin: 0 auto;
              text-align: center;
              border-color: red;
              padding-bottom: 15px;
              margin-bottom: 30px;
            `}
            size={15}
            color="#f9d342"
            loading={loading}
          />
        ) : (
          ''
        )}
        <ul className="list-block comments-wrapper">
          {!loading
            && (comments || []).map((comment, key) => (
              <li key={key}>
                {comment && comment ? (
                  <div>
                    {/* avatar */}
                    <div className="comment-icon hide-on-small">
                      <div className="image">
                        <Img
                          imgSrc={(comment.commentAuthor && comment.commentAuthor.image) || avatar}
                          alt="Profile"
                        />
                      </div>
                    </div>
                    {/* comment body */}
                    <div className="comment-message-wrapper">
                      <div>
                        <Link to="/" className="text-success">
                          {`${comment.commentAuthor.firstName} ${comment.commentAuthor.lastName}`}{' '}
                          {comment.commentAuthor.username || ''}
                        </Link>
                        <br />
                        <div className="text-grey small-v-margin timestamp">
                          {timeStamp(comment.createdAt)}
                        </div>
                      </div>
                      <div className="large-v-padding">{comment.body}</div>
                      {/* comment actions */}
                      <div className="row">
                        <span className="small-screen-4 medium-screen-4 large-screen-2">
                          <button className="light text-black">
                            <FontAwesomeIcon icon={faThumbsUp} className="text-grey" />
                            &nbsp; likes
                          </button>
                        </span>
                        {isAuth && profile.id === comment.userId ? (
                          <span className="small-screen-4 medium-screen-4 large-screen-2 right-align">
                            <button className="light text-black">
                              <FontAwesomeIcon icon={faClock} className="text-grey" />
                              &nbsp; Edit History
                            </button>
                            <button
                              onClick={() => this.toggleCommentEditor(key)}
                              className="success text-white"
                            >
                              <FontAwesomeIcon icon={faPenAlt} />
                              &nbsp; Edit
                            </button>
                            <button
                              onClick={() => {
                                // eslint-disable-next-line no-alert
                                if (window.confirm('Are you sure you want to delete this comment?')) this.onDeleteComment(comment.id, key);
                              }}
                              className="danger text-white"
                            >
                              <FontAwesomeIcon icon={faTrash} />
                              &nbsp; Delete
                            </button>
                          </span>
                        ) : (
                          ''
                        )}
                      </div>
                      <div className="divider white" />
                      <div
                        className={`comment-editor ${(commentEditor[key]
                          && commentEditor[key].display)
                          || 'hide'}`}
                      >
                        <form onSubmit={e => this.onSubmit(e, key, comment.id, comment.userId)}>
                          <div>Edit this comment</div>
                          <div className="input-field">
                            <TextareaAutosize
                              name={`comment-${key}`}
                              placeholder="Edit your comment here..."
                              onChange={this.onChangeEditComment}
                              value={
                                (newComments[`comment-${key}`]
                                  && newComments[`comment-${key}`].value)
                                || comment.body
                              }
                              id={`edit-comment-${key}`}
                              style={{ minHeight: 12, fontSize: 14 }}
                              className="shadow-1"
                            />
                          </div>
                          <div className="input-field">
                            <button
                              onSubmit={this.onSubmit}
                              type="submit"
                              className="button success text-white radius-5"
                            >
                              Update
                            </button>
                            <button
                              onClick={e => this.closeCommentEditor(e, key)}
                              className="button white text-black radius-5 border b-grey"
                            >
                              Cancel
                            </button>
                            {loading ? (
                              <span className="medium-padding radius-5">loading</span>
                            ) : (
                              ''
                            )}
                            {errors && errors.message ? (
                              <span className="medium-padding radius-5 text-danger">
                                {errors.message}
                              </span>
                            ) : (
                              ''
                            )}
                          </div>
                        </form>
                      </div>
                      <div className="divider white" />
                    </div>
                    <div className="divider white" />
                  </div>
                ) : (
                  ''
                )}
              </li>
            ))}
          <div>
            {!loading && comments && !comments.length > 0 ? (
              <div className="center-align">
                <FontAwesomeIcon
                  icon={faCommentAlt}
                  className="text-light-grey fa-4x center-align"
                />
                <h2 className="center-align large-v-padding">No comment yet</h2>
              </div>
            ) : (
              ''
            )}
          </div>
        </ul>
        <div className="clear" />
      </div>
    );
  }
}

CommentThread.propTypes = {
  fetchComments: PropTypes.func.isRequired,
  deleteComment: PropTypes.func.isRequired,
  editComment: PropTypes.func.isRequired,
  isAuth: PropTypes.bool,
  profile: PropTypes.object,
  comments: PropTypes.array,
  comment: PropTypes.object,
  slug: PropTypes.string,
  rating: PropTypes.number,
  params: PropTypes.object,
  match: PropTypes.object,
  loading: PropTypes.bool,
  deleted: PropTypes.any,
  message: PropTypes.bool
};
const mapStateToProps = ({
  user: { isAuth, profile },
  comments: {
    deleteComment: { message, deleted },
    fetchComments: { comments, errors }
  }
}) => ({ isAuth, profile, comments, message, errors, deleted });

export default connect(
  mapStateToProps,
  { fetchComments, deleteComment, editComment }
)(CommentThread);
