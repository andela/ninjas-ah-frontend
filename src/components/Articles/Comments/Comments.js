import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CommentThread from './CommentThread';
import CommentForm from './CommentForm';
import './Comments.scss';

export default class Comments extends Component {
  state = { title: 'Comments' };

  render() {
    return (
      <div id="wrap-comments" className="radius-2">
        <h1 className="nobold medium-h-padding medium-v-padding center-align">
          {this.state.title}
        </h1>
        <CommentForm slug={this.props.slug} />
        <div className="clear" />
        <CommentThread slug={this.props.slug} />
        <div className="clear" />
      </div>
    );
  }
}

Comments.propTypes = { slug: PropTypes.string };
