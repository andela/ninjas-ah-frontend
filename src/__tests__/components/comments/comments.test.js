import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'react-thunk';
import { Provider } from 'react-redux';
import article from '../../../__mocks__/article';
import { CommentForm as CommentFormComponent } from '../../../components/Articles/Comments/CommentForm';
import { shallow, mount } from '../../../../config/enzymeConfig';

describe('<CommentComponent />', () => {
  const props = {
    errors: { error: ['12'] },
    comments: [
      {
        body: 'Hello John Doe',
        userId: 1,
        commentAuthor: {
          firstName: 'John',
          lastName: 'Doe',
          username: 'johndoe',
          image: 'image.png'
        },
        createdAt: '2019-01-01'
      }
    ],
    createComment: jest.fn()
  };
  const state = { title: 'comments' };
  const component = shallow(<CommentFormComponent {...props} />);
  it('should render a <CommentFormComponent /> component ', () => {
    expect(component).toMatchSnapshot();
  });
  it('should set state', () => {
    component.setState({ ...state });
    expect(component).toHaveLength(1);
  });
  it('should call onChange method when the comment is provided', () => {
    const component = shallow(<CommentFormComponent {...props} />);
    const spy = jest.spyOn(component.instance(), 'onChange');
    component.instance().forceUpdate();
    component.setProps({ ...props, errors: { errors: ['error1'] } });
    const event = { target: { value: 'body' } };
    const textArea = component.find('#commentBody');
    textArea.simulate('change', event);
    expect(textArea.length).toBe(1);
    expect(spy).toHaveBeenCalled();
  });
  it('should submit comment', () => {
    const instance = component.instance();
    const fakeEvent = { preventDefault: () => {} };
    instance.onSubmit(fakeEvent);
  });
  it('should not comment if not logged in ', () => {
    component.setProps({ isAuth: false });
    component.setState({ errors: { token: 'some' } });
    const fakeEvent = { preventDefault: () => {} };
});
