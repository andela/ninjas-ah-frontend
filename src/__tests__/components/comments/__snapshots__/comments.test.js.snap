// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentComponent /> should render a <CommentFormComponent /> component  1`] = `ShallowWrapper {}`;

exports[`<CommentFormComponent /> should render a <CreateCommentFormComponent /> component  1`] = `ShallowWrapper {}`;
